src/coreclr/jit/layout.cpp

```diff
@@ -722,6 +784,13 @@ bool ClassLayout::AreCompatible(const ClassLayout* layout1, const ClassLayout* l
     {
         if (layout1->GetGCPtrType(i) != layout2->GetGCPtrType(i))
         {
+            // Allow a source GC_REF to match a custom GC_BYREF
+            //
+            if ((layout2->GetGCPtrType(i) == TYP_REF) && (layout1->GetGCPtrType(i) == TYP_BYREF) &&
+                layout1->IsCustomLayout())
+            {
+                continue;
+            }
```

Comment:
Changing the definition of layout compatibility seems quite questionable to me. Why is it necessary?

Reply:
```
GCStruct a;
a.obj = ..;   // doesn't escape

GCStruct b;
b.obj = ..;   // escapes

// Assignment is legal, but layouts are not compatible w/o the above
// a.obj must be BYREF as it may refer to stack or heap

a = b
```

Reply:
I was tempted to retain the class handle in new layout, but didn't.

Reply:
Note allowing `b = a` would be incompatible with the results of escape analysis, which is why this is one-sided (source can be byref when dest is ref, but not vice-versa).

Reply:
So are you saying that for the
```csharp
object aObj = ...; // doesn't escape

object bObj = ...; // escapes

aObj = bObj;
```
case, today we will produce a `STORE_LCL_VAR` that stores a `TYP_REF` into a `TYP_BYREF`?

Reply:
Yes, we will retype `aObj` appearances to be byrefs, as it will be "possibly stack pointing".

Reply:
Ok, it makes it somewhat more palatable. But it seems like we need to separate the notions of "user facing" layout compatibility, as used by e.g. `Unsafe.BitCast`, and "implementation-specific assignment compatibility" which is non-reflexive and where we rely on internal coreclr implementation details. As you probably know there could be a potential future where byrefs pointing at the method table would be illegal... I guess in that case we would need a third kind of GC pointer. Or we would have to switch object stack allocation to only kick in when it can reason about and change all accesses of the object.

Reply:
I would just suggest to rename this to `ClassLayout::AssignmentCompatible(ClassLayout* dest, ClassLayout* src)` and allow the byref <- ref assignments generally. And then switch `Unsafe.BitCast` to a version that keeps the stricter behavior.
Should probably be a separate PR...

Reply:
(Maybe `BitCast` doesn't even need any changes, I was under the assumption that it threw an exception on incompatible layouts, but it does not seem like it is the case)

Reply:
Ok, let me split out the existing uses in a separate PR.

I need to look at what escape analysis will do for struct retyping that preserves GCness to see if we ever end up with a case like

```
GCStruct    a;   // does not escape
GCStructAlt b;   // escapes,

b = Unsafe.BitCast(a);
```
and make sure we know to retype `b`.